### PR TITLE
Multimodality/sft tool masking

### PR DESCRIPTION
**Pull request by dtamayo-nlp**
## Motivation

Tool outputs are embedded directly inside assistant messages, between <|tools_suffix|> and the next <|tools_prefix|>:
```
...<|assistant_start|>I'll help you find rhyming options for your birthday poem! Let me start by getting rhyming words for "smile" and counting the syllables in your current line.<|tools_prefix|>[{"lyrical-mcp-find_rhymes": {"input_word": "smile"}}]<|tools_suffix|>[{
  "1_syllable": [
    "aisle",
    "bile",
    "bille",
    "cheil",
    "deihl",
    "dial",
    "file",
    "geil",
    "gile",
    "gille",
    "guile",
    "heil",
    "hile",
    "hyle",
    "i'll",
    "isle",
    "kile",
    "kyl",
    "kyle",
    "lile"
  ],
  "2_syllable": [
    "argyll",
    "awhile",
    "beguile",
    "compile",
    "delisle",
    "fertile",
    "hostile",
    "kurzweil",
    "marseille",
    "mikhail",
    "monteil",
    "nevile",
    "panfile",
    "refile",
    "restyle",
    "revile",
    "soleil",
    "vantuyl",
    "worthwhile"
  ],
  "3_syllable": [
    "versatile"
  ]
}]<|tools_prefix|>[{"lyrical-mcp-count_syllables": {"input_string": "With joy that leaps from every smile"}}]<|tools_suffix|>[9]Great! Here's what I found:

## Rhyming Options for "smile":
- **1 syllable**: aisle, file, guile, isle, kyle, tile, while
- **2 syllables**: awhile, beguile, compile, fertile, hostile, worthwhile
- **3 syllables**: versatile
...
```
Previously, loss masking only needed to unmask everything between `<|assistant_start|>` and `<|assistant_end|>`. Tool outputs now sit inside that span and must be excluded from the training loss, otherwise the model is trained to predict deterministic tool responses it has no control over.

## Problem

Bracket matching on raw token IDs is not straightforward: `[` and `]` appear in hundreds of compound tokens (`[{`, `"}]`, `"]["`, `],` ...). A naïve check against the single-character token ID silently misses all of them, producing incorrect bracket depth and thus wrong mask boundaries.

## Implementation

**1. Bracket delta lookup table** (`__init__`): a tensor of shape `(vocab_size,)` is built once at dataset initialisation. For every token in the vocabulary, its entry stores the net bracket contribution (`count('[') - count(']')`), covering all compound forms correctly.

**2. Cumsum bracket matching** (`_get_ltor_masks_and_position_ids`): for each `<|tools_suffix|>` position, a depth-tracking cumulative sum over the lookup values is computed once for the full sequence in O(N). The closing `]` is then found with a single `torch.where` call per tool call rather than a Python loop over tokens. Both `loss_mask` and `assistant_mask` are zeroed over the matched span.

### CODE REVIEW

**Review by RaphaelKreft — COMMENTED** [review]
# SFT Tool Output Masking — Analysis

Branch: `multimodality/sft-tool-masking`
File: `megatron/training/datasets/apertus_sft_dataset.py`

## How It Works

The chat template wraps every tool interaction as:

```
<|tools_prefix|>[{"tool": args}]<|tools_suffix|>[tool output]
```

The algorithm finds each `<|tools_suffix|>`, then bracket-matches the `[...]`
that follows using a per-token delta lookup (`count('[') - count(']')` per
vocab entry, handling compound tokens like `[{`, `}]`). It masks from the
opening `[` to where cumulative depth returns to baseline.

## Case 1 — Balanced / No Brackets

Tool output is valid JSON, has balanced brackets, or no brackets at all.
The template's wrapping `[...]` ensures the cumsum returns to baseline at
exactly the right position. **All tool output is masked, no assistant tokens
affected.** This is the expected common case.

## Case 2 — Unbalanced Brackets

When the raw tool return contains stray `[` or `]` characters, the bracket
invariant breaks:

| Content issue          | Failure        | Effect                                  | Severity |
|------------------------|----------------|-----------------------------------------|----------|
| Extra `]`              | Under-mask     | Tail of tool output keeps loss=1        | Low      |
| Extra `[`, later `]`   | **Over-mask**  | Assistant tokens after output lose loss  | **High** |
| Extra `[`, no later `]`| No mask at all | Entire tool output keeps loss=1         | Medium   |

Over-masking is the most dangerous: the search bleeds past the tool output
into assistant text, silently dropping gradient on real tokens.

## Case 3 — Compound Token Order Sensitivity

The delta lookup stores the **net** bracket count per token. A token like `][`
has `delta=0`, hiding the fact that it closes before it opens. The algorithm
cannot distinguish it from a bracketless token.

Ex. This can happen if we have a closing bracket of tool response and then an opening one directly after as part of a assistant response.

The token `][` exists in the tokenizer (ID 5371, `net=0`). Two other
order-sensitive tokens also exist: `]][` (ID 95608, `net=-1`) and ` [[]`
(ID 35976, `net=+1`). These are unlikely to appear in well-formed JSON tool
output but could surface in free-text or malformed responses.

## Mitigation

**Preferred:** Add two dedicated special tokens (e.g. `<|tool_output_start|>`
/ `<|tool_output_end|>`) to explicitly delimit tool responses. This makes the
masking logic a simple `get_matching_mask_by_start_end()` call — clean, fast,
and immune to content edge cases.

**Modest alternative:** Clamp the bracket search to the next `<|tools_prefix|>`
or `<|assistant_end|>` as an upper bound. This prevents over-masking into
assistant text in multi-turn conversations while keeping the current bracket
approach.

**Comment by RaphaelKreft** [comment]
So now it depends on how we want to proceed: If we are okay with the failure cases we can proceed. As for me, I would be in favor of having separate special tokens to prevent the issues regarding the failure cases now and for future.

**Comment by ischlag** [comment]
Can we just drop the failure cases? I'd assume they are a small minority.

**Comment by RaphaelKreft** [comment]
> Can we just drop the failure cases? I'd assume they are a small minority.

I have not counted the potential cases in our data. Though any sample with 1) an uneven number of brackets 2) compound brackets "][" as part of the tool respone can be an issue. Also if the model response after tool call starts with [.

I can count samples affected - but I think for now and future purposes its faster and cleaner to have separate special tokens. (Masking based on special tokens is trivial to implement and I think @dtamayo-nlp is already on it)

**Comment by dtamayo-nlp** [comment]
I agree. Give us two hours, Ill have the tokenized data with the new chat template ready in about 20 minutes. The chat template has already passed the tests. I just need to implement the masking in the scripts (which should be quick) and then run a thorough end-to-end validation.

**Comment by dtamayo-nlp** [comment]
Thanks a lot to @RaphaelKreft for detecting the issues. I've force-pushed an update to this branch with a simpler implementation. It should no longer have any edge cases. I'll also document the tokenization scripts using the new chat template in the GitHub issue and in the main channel in a few mins, since this may affect some RL users.

Edit: It needs having the data tokenized with `<|tool_output_start|>` / `<|tool_output_end|>`
